cmd/signing-server/main.go

```diff
@@ -69,14 +73,126 @@ type Config struct {
 	DisableAuth        bool
 	DisableHTTPS       bool
 
+	HSMModule   string
+	HSMSlot     int
+	HSMPass     string
+	HSMKeyLabel string
+	HSMKeyId    string
+
+	HSMContext   *pkcs11.Ctx
+	HSMSession   pkcs11.SessionHandle
+	HSMKeyHandle pkcs11.ObjectHandle
+
 	// calculated by program
 	Logger *zap.Logger
 }
 
+func (c *Config) SetupHSM() error {
+	c.Logger.Info("setup HSM signing", zap.String("module", c.HSMModule))
+	p := pkcs11.New(c.HSMModule)
+	err := p.Initialize()
+	if err != nil {
+		return err
+	}
+
+	if c.HSMSlot < 0 {
+		slots, err := p.GetSlotList(true)
+		if err != nil {
+			return fmt.Errorf("lookup HSM slots: %w", err)
+		} else {
+			c.HSMSlot = int(slots[0])
+		}
+	}
+	c.Logger.Info("using slot", zap.Int("slot", c.HSMSlot))
+
+	session, err := p.OpenSession(uint(c.HSMSlot), pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	if err != nil {
+		return fmt.Errorf("open HSM session: %w", err)
+	}
+
+	err = p.Login(session, pkcs11.CKU_USER, c.HSMPass)
+	if err != nil {
+		p.CloseSession(session)
+		return fmt.Errorf("HSM login: %w", err)
+	}
+
+	attrs := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+	}
+	if c.HSMKeyLabel != "" {
+		attrs = append(attrs, pkcs11.NewAttribute(pkcs11.CKA_LABEL, c.HSMKeyLabel))
+	}
+	if c.HSMKeyId != "" {
+		id, err := strconv.ParseInt(c.HSMKeyId, 10, 8)
+		if err != nil {
+			p.CloseSession(session)
+			p.Destroy()
+			return fmt.Errorf("invalid key id %q: %w", c.HSMKeyId, err)
+		}
+		attrs = append(attrs, pkcs11.NewAttribute(pkcs11.CKA_ID, []byte{byte(id)}))
+	}
+	if err := p.FindObjectsInit(session, attrs); err != nil {
+		p.CloseSession(session)
+		p.Destroy()
+		return fmt.Errorf("HSM get private key handle %q: %w", c.HSMKeyId, err)
+	}
+
+	objs, _, err := p.FindObjects(session, 1)
+	if err != nil {
+		p.CloseSession(session)
+		p.Destroy()
+		return fmt.Errorf("key id %q: %w", c.HSMKeyId, err)
+	}
+	if len(objs) != 1 {
+		p.CloseSession(session)
+		p.Destroy()
+		return fmt.Errorf("key id %q not found (%d)", c.HSMKeyId, len(objs))
+	}
+	p.FindObjectsFinal(session)
+
+	c.Logger.Info("key id", zap.Uint("id", uint(objs[0])))
+
+	c.HSMSession = session
+	c.HSMContext = p
+	c.HSMKeyHandle = objs[0]
+	return nil
+}
+
 func (c *Config) Validate(args []string) error {
-	if c.SigningPrivateKeyPath == "" {
-		return errors.New("path to private key file must be set")
+	if c.SigningPrivateKeyPath == "" && c.HSMModule == "" {
```

Comment:
This isn't right here with the error below. It's not an OR. It's an AND. So if neither of them are set, this will fail.
It will only ever pass if BOTH are set.